core/domain/entities/routing/handlers/admin/EspressoEventEditor.php

```diff
@@ -0,0 +1,133 @@
+                'EventEspresso\core\domain\services\admin\events\editor\EventEntityRelations'    => EE_Dependency_Map::load_from_cache,
+            ]
+        );
+        $this->dependency_map->registerDependencies(
+            'EventEspresso\core\domain\services\admin\events\editor\EventEntityRelations',
+            [
+                'EEM_Datetime'   => EE_Dependency_Map::load_from_cache,
+                'EEM_Event'      => EE_Dependency_Map::load_from_cache,
+                'EEM_Price'      => EE_Dependency_Map::load_from_cache,
+                'EEM_Price_Type' => EE_Dependency_Map::load_from_cache,
+                'EEM_Ticket'     => EE_Dependency_Map::load_from_cache,
+            ]
+        );
+        $this->dependency_map->registerDependencies(
+            'EventEspresso\core\domain\services\admin\events\editor\NewEventDefaultEntities',
+            [
+                'EEM_Datetime'                                                       => EE_Dependency_Map::load_from_cache,
+                'EEM_Event'                                                          => EE_Dependency_Map::load_from_cache,
+                'EEM_Price'                                                          => EE_Dependency_Map::load_from_cache,
+                'EEM_Price_Type'                                                     => EE_Dependency_Map::load_from_cache,
+                'EEM_Ticket'                                                         => EE_Dependency_Map::load_from_cache,
+                'EventEspresso\core\domain\services\admin\entities\DefaultDatetimes' => EE_Dependency_Map::load_from_cache,
+            ]
+        );
+        $this->dependency_map->registerDependencies(
+            'EventEspresso\core\domain\services\admin\entities\DefaultDatetimes',
+            [
+                'EventEspresso\core\domain\services\admin\entities\DefaultTickets' => EE_Dependency_Map::load_from_cache,
+                'EEM_Datetime'                                                     => EE_Dependency_Map::load_from_cache,
+            ]
+        );
+        $this->dependency_map->registerDependencies(
+            'EventEspresso\core\domain\services\admin\entities\DefaultTickets',
+            [
+                'EventEspresso\core\domain\services\admin\entities\DefaultPrices' => EE_Dependency_Map::load_from_cache,
+                'EEM_Ticket'                                                      => EE_Dependency_Map::load_from_cache,
+            ]
+        );
+        $this->dependency_map->registerDependencies(
+            'EventEspresso\core\domain\services\admin\entities\DefaultPrices',
+            [
+                'EEM_Price'      => EE_Dependency_Map::load_from_cache,
+                'EEM_Price_Type' => EE_Dependency_Map::load_from_cache,
+            ]
+        );
+    }
+
+
+    /**
+     * implements logic required to run during request
```

Comment:
it's a middleware, kinda?

Reply:
mmm... no it's not middleware it's more like undefined logic that's been spread out all over the place that's been extracted and moved into a formally defined interface that can be better managed by the system.

Previously the request detection and subsequent handling of said requests were being done directly inside `EE_System` (and other places) with a ton of other logic all mixed up together. Dependencies for any classes being loaded were just ALL hardcoded directly in `EE_Dependency_Map` regardless of whether or not they were actually needed for the current request.

So now, that logic has been moved into these `Route` classes, which currently only have the following three methods for their public interface:

- `matchesCurrentRequest()` returns true if the current request matches this route

- `handleRequest()` logic to run when above method returns `true`

- `isHandled()` true if `handleRequest()` has already run

and the new `RouteHandler` class can execute that logic for any incoming request that has been added to its internal collection which is done simply by providing the FQCN to the Route class:

```php
 $this->router->addRoute('EventEspresso\core\domain\entities\routing\handlers\admin\EspressoEventEditor');
```

So all of the previous logic inside that Route class which was previously in `EE_System` has been replaced by that one line of code in `EE_System`, and the dependencies that previously were ALWAYS loaded in `EE_Dependency_Map` are now only registered if they are needed for the current request.

Reply:
thank you @tn3rb for such an elaborate explanation :bow: 

it's interesting to understand some of the underlying server-side logic, not just the client-side of things, I never know from where I can get some inspiration for some patterns